### PR TITLE
add foreign key on update and on delete support

### PR DESCRIPTION
**Pull request by zxylvlp**

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -659,6 +661,16 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 
 	// Set empty constraint names.
 	for _, constr := range constraints {
+		if constr.Tp == ast.ConstraintForeignKey {
+			if constr.Name == "" {
+				i, err := d.genGlobalID()
```
Comment (coocood) [inline ddl/ddl.go]: Why not generate a name as following?

File: ddl/ddl.go [hunk 2]
```diff
@@ -687,6 +699,33 @@ func (d *ddl) buildTableInfo(tableName model.CIStr, cols []*column.Col, constrai
 		tbInfo.Columns = append(tbInfo.Columns, &v.ColumnInfo)
 	}
 	for _, constr := range constraints {
+		if constr.Tp == ast.ConstraintForeignKey {
+			for _, fk := range tbInfo.ForeignKeys {
```
Comment (coocood) [inline ddl/ddl.go]: The `constr.Name` is uniquely generated, so this check is not needed.

Comment (zxylvlp) [inline ddl/ddl.go]: It may be assigned by user.

File: column/column.go [hunk 3]
```diff
@@ -42,6 +42,11 @@ type IndexedCol struct {
 	X kv.Index
 }
 
+// ForeignKey be used to wrap a foreign key info.
+type ForeignKey struct {
```
Comment (zxylvlp) [inline column/column.go]: It is more unified in this way, and it can be delete. Should I remove it?

File: table/table.go [hunk 4]
```diff
@@ -50,7 +50,7 @@ type Table interface {
 	Indices() []*column.IndexedCol
 
 	// ForeignKeys returns the foreign keys of the table.
-	ForeignKeys() []*column.ForeignKey
+	ForeignKeys() []*model.FKInfo
```
Comment (coocood) [inline table/table.go]: We don't need this method, we can get FKInfo in `Meta`.

**Comment by zxylvlp** [comment]
@coocood PTAL

File: optimizer/resolver.go [hunk 5]
```diff
@@ -100,6 +100,8 @@ type resolverContext struct {
 	inCreateOrDropTable bool
 	// When visiting show statement.
 	inShow bool
+	// When visiting constraint foreign key.
+	inConstraintForeignKey bool
```
Comment (coocood) [inline optimizer/resolver.go]: When table does not exist, foreign key should fail to resolve, so this is not needed.

**Comment by zxylvlp** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@zimulala PTAL

**Comment by qiuyesuifeng** [comment]
Seems we are going to support foreign key now? If that we should add more and more tests to cover it.
In most internet companies, it is not suggested to use foreign key in online service.

**Comment by shenli** [comment]
@qiuyesuifeng Not really. We just support some foreign key related syntax.

File: parser/parser.y [hunk 6]
```diff
@@ -1077,9 +1087,58 @@ ConstraintElem:
...
 ReferDef:
-	"REFERENCES" TableName '(' IndexColNameList ')'
+	"REFERENCES" TableName '(' IndexColNameList ')' OnDeleteOpt OnUpdateOpt
+	{
+		var onDeleteOpt *ast.OnDeleteOpt
+		if $6 != nil {
+			onDeleteOpt = $6.(*ast.OnDeleteOpt)
+		}
+		var onUpdateOpt *ast.OnUpdateOpt
+		if $6 != nil {
```
Comment (zimulala) [inline parser/parser.y]: ` $6 != nil` -> ` $7 != nil` ？

File: model/model.go [hunk 7]
```diff
@@ -162,6 +169,34 @@ func (index *IndexInfo) Clone() *IndexInfo {
...
+	Name     CIStr       `json:"fk_name"`
+	RefTable CIStr       `json:"ref_table"`
+	RefCols  []CIStr     `json:"ref_cols"`
+	Cols     []CIStr     `json:"cols"`
+	OnDelete int         `json:"on_delete"`
+	OnUpdate int         `json:"on_update"`
+	State    SchemaState `json:"state"`
+}
+
+// Clone clones FKInfo
```
Comment (zimulala) [inline model/model.go]: Add . at the end of a line.

File: parser/parser.y [hunk 8]
```diff
@@ -1077,9 +1087,58 @@ ConstraintElem:
...
+			Table: $2.(*ast.TableName),
+			IndexColNames: $4.([]*ast.IndexColName),
+			OnDelete: onDeleteOpt,
+			OnUpdate: onUpdateOpt,
+		}
+	}
+
+OnDeleteOpt:
+	{
+		$$ = &ast.OnDeleteOpt{}
```
Comment (zimulala) [inline parser/parser.y]: `$$ = nil` is better ?

File: infoschema/infoschema.go [hunk 9]
```diff
@@ -456,6 +461,10 @@ const (
 	codeTableNotExists                   = 1146
 	codeColumnNotExists                  = 1054
 
+	codeForeignKeyNotMatch  = 1055
```
Comment (zimulala) [inline infoschema/infoschema.go]: The error code is conflict with MySQL error code.

File: model/model.go [hunk 10]
```diff
@@ -162,6 +169,34 @@ func (index *IndexInfo) Clone() *IndexInfo {
...
+func (fk *FKInfo) Clone() *FKInfo {
+	nfk := *fk
+
+	nfk.RefCols = make([]CIStr, len(fk.RefCols))
+	nfk.Cols = make([]CIStr, len(fk.Cols))
+	for i := range fk.RefCols {
```
Comment (zimulala) [inline model/model.go]: Using `copy(nfk.RefCols, fk.RefCols)` ?

File: ddl/ddl.go [hunk 11]
```diff
@@ -1005,6 +1059,60 @@ func (d *ddl) CreateIndex(ctx context.Context, ti ast.Ident, unique bool, indexN
 	return errors.Trace(err)
 }
 
+func (d *ddl) CreateForeignKey(ctx context.Context, ti ast.Ident, fkName model.CIStr, keys []*ast.IndexColName, refer *ast.ReferenceDef) error {
+	is := d.infoHandle.Get()
+	schema, ok := is.SchemaByName(ti.Schema)
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.Gen("database %s not exists", ti.Schema)
```
Comment (zimulala) [inline ddl/ddl.go]: Unify the error information in line 1066 and line 1096 ?

File: ddl/foreign_key.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+		fkInfo   model.FKInfo
+		fkName   model.CIStr
+		fkID     int64
+		keys     []*ast.IndexColName
+		refTable *ast.TableName
+		refKeys  []*ast.IndexColName
+		onDelete ast.ReferOptionType
+		onUpdate ast.ReferOptionType
+	)
+	err = job.DecodeArgs(&fkName, &fkID, &keys, &refTable, &refKeys, &onDelete, &onUpdate)
```
Comment (zimulala) [inline ddl/foreign_key.go]: Why not use `job.DecodeArgs(&fkInfo)` directly？

File: ddl/foreign_key.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+		return nil
+	default:
+		return ErrInvalidForeignKeyState.Gen("invalid fk state %v", fkInfo.State)
+	}
+}
+
+func (d *ddl) onDropForeignKey(t *meta.Meta, job *model.Job) error {
+	schemaID := job.SchemaID
+	tblInfo, err := d.getTableInfo(t, job)
+
```
Comment (zimulala) [inline ddl/foreign_key.go]: Remove this empty line.

File: ddl/foreign_key.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+		return infoschema.ErrForeignKeyNotExists.Gen("foreign key doesn't exist", fkName)
+	}
+
+	nfks := tblInfo.ForeignKeys[:0]
+	for _, fk := range tblInfo.ForeignKeys {
+		if fk.Name.L != fkName.L {
+			nfks = append(nfks, fk)
+		}
+	}
+
```
Comment (zimulala) [inline ddl/foreign_key.go]: Remove this empty line.

File: ddl/foreign_key.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	tblInfo.ForeignKeys = append(tblInfo.ForeignKeys, &fkInfo)
+
+	_, err = t.GenSchemaVersion()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	switch fkInfo.State {
+	case model.StateNone:
+		// none -> public
```
Comment (zimulala) [inline ddl/foreign_key.go]: Add more comments.

**Comment by zxylvlp** [comment]
@zimulala PTAL

File: ddl/ddl.go [hunk 16]
```diff
@@ -1059,6 +1067,34 @@ func (d *ddl) CreateIndex(ctx context.Context, ti ast.Ident, unique bool, indexN
 	return errors.Trace(err)
 }
 
+func (d *ddl) buildFkInfo(fkName model.CIStr, keys []*ast.IndexColName, refer *ast.ReferenceDef) (*model.FKInfo, error) {
```
Comment (zimulala) [inline ddl/ddl.go]: buildFkInfo -> buildFKInfo ?

**Comment by zimulala** [comment]
Rest LGTM

**Pull request merged**